peps/pep-0778.rst

```diff
@@ -0,0 +1,357 @@
+debuggers, we can leverage the interpreter's existing evaluation loop and thread
+state to coordinate debugging operations.
+
+The mechanism works by having debuggers write to specific memory locations in
+the target process that the interpreter then checks during its normal execution
+cycle. When the interpreter detects that a debugger wants to attach, it executes the
+requested operations only when it's safe to do so - that is, when no internal
+locks are held and all data structures are in a consistent state.
+
+
+Runtime State Extensions
+------------------------
+
+A new structure is added to PyThreadState to support remote debugging:
+
+.. code-block:: C
+
+    typedef struct _remote_debugger_support {
+        int debugger_pending_call;
+        char debugger_script[MAX_SCRIPT_SIZE];
+    } _PyRemoteDebuggerSupport;
+
+
+This structure is appended to ``PyThreadState``, adding only a few fields that
+are **never accessed during normal execution**. The ``debugger_pending_call`` field
+indicates when a debugger has requested execution, while ``debugger_script``
+provides Python code to be executed when the interpreter reaches a safe point.
+
+
+Debug Offsets Table
+-------------------
+
+
+Python 3.12 introduced a debug offsets table placed at the start of the
+PyRuntime structure. This section contains the ``_Py_DebugOffsets`` structure that
+allows external tools to reliably find critical runtime structures regardless of
+`ASLR <https://en.wikipedia.org/wiki/Address_space_layout_randomization>`__ or
+how Python was compiled.
+
+This proposal extends the existing debug offsets table with new fields for
+debugger support:
+
+.. code-block:: C
+
+    struct _debugger_support {
+        uint64_t eval_breaker;            // Location of the eval breaker flag
+        uint64_t remote_debugger_support; // Offset to our support structure
+        uint64_t debugger_pending_call;   // Where to write the pending flag
+        uint64_t debugger_script;         // Where to write the script
+    } debugger_support;
+
+These offsets allow debuggers to locate critical debugging control structures in
+the target process's memory space. The ``eval_breaker`` and ``remote_debugger_support``
+offsets are relative to each ``PyThreadState``, while the ``debugger_pending_call``
+and ``debugger_script`` offsets are relative to each ``_PyRemoteDebuggerSupport``
+structure, allowing the new structure and its fields to be found regardless of
+where they are in memory.
+
+Attachment Protocol
+-------------------
+When a debugger wants to attach to a Python process, it follows these steps:
+
+1. Locate ``PyRuntime`` structure in the process:
+   - Find Python binary (executable or libpython) in process memory (OS dependent process)
+   - Extract ``.PyRuntime`` section offset from binary's format (ELF/Mach-O/PE)
+   - Calculate the actual ``PyRuntime`` address in the running process by relocating the offset to the binary's load address
```

Comment:
I think we need a newline here for the sublist to render as a list and not a paragraph:
```suggestion
1. Locate ``PyRuntime`` structure in the process:

   - Find Python binary (executable or libpython) in process memory (OS dependent process)
   - Extract ``.PyRuntime`` section offset from binary's format (ELF/Mach-O/PE)
   - Calculate the actual ``PyRuntime`` address in the running process by relocating the offset to the binary's load address
```